specifyweb/frontend/js_src/lib/components/AttachmentsBulkImport/utils.ts

```diff
@@ -426,7 +443,93 @@ export function resolveAttachmentStatus(
 export const getAttachmentsFromResource = (
   baseResource: SerializedResource<Tables['CollectionObject']>,
   relationshipName: string
+): {
+  readonly key: keyof SerializedResource<Tables['CollectionObject']>;
+  readonly values: RA<SerializedResource<Tables['CollectionObjectAttachment']>>;
+} =>
+  defined(
+    mappedFind(Object.entries(baseResource), ([key, value]) =>
+      key.toLowerCase() === relationshipName.toLowerCase()
+        ? {
+            key,
+            values: value as RA<
+              SerializedResource<Tables['CollectionObjectAttachment']>
+            >,
+          }
+        : undefined
+    )
+  );
+
+type RecordResponse =
+  | State<
+      'invalid',
+      { readonly reason: keyof typeof keyLocalizationMapAttachment }
+    >
+  | State<'valid', { readonly record: SerializedResource<CollectionObject> }>;
+
+const wrapAjaxRecordResponse = async (
+  ajaxPromise: () => Promise<
+    AjaxResponseObject<SerializedModel<CollectionObject>>
+  >,
+  // Defines errors on which to not trigger a retry.
+  statusMap: RR<number | 'fallback', keyof typeof keyLocalizationMapAttachment>,
+  triggerRetry?: () => void
+): Promise<RecordResponse> =>
+  ajaxPromise().then(({ data, status }) => {
+    if (statusMap[status] !== undefined)
+      return { type: 'invalid', reason: statusMap[status] };
+    if (status !== Http.OK) {
+      triggerRetry?.();
+      return { type: 'invalid', reason: statusMap.fallback };
+    }
+    return {
+      type: 'valid',
+      record: serializeResource(data),
+    };
+  });
+
+const baseStatusMap = {
+  [Http.NOT_FOUND]: 'nothingFound',
+  [Http.CONFLICT]: 'saveConflict',
+} as const;
+
+export const fetchForAttachmentUpload = async (
+  baseTableName: keyof Tables,
+  matchId: number,
+  triggerRetry?: () => void
```

Comment:
Good idea, but it will be used in two places anyways (for fetching, and saving), and then each time for upload and rollback, so it would need to be generalized (so the triggerRetry will be passed down) regardless